### PR TITLE
Shell optimizations; reduce subshell/system calls; reduce repeated code

### PR DESCRIPTION
**Pull request by zpiatt**
Now that there's a baseline for linting, I focused on optimization. There's a lot more that can be done, but this is about all I have time for this week. I made some assumptions based on your prior feedback, and the overall flow of the scripts, but I have no problem reverting anything you don't like.

1. Variables assigned lowercase with `echo <var> | awk '{print tolower($0)}'` have been changed to `${var,,}`.
2. I removed `-exec` from the `find` commands. The `exec` function was written poorly. At best it's slightly slower than piping into `xargs`, at worst it's significantly slower.
3. Every `return` and `exit` that was otherwise unspecified ( implicit `0` ) is now an explicit `0` ( I suspect you may want some of these to be `1` ).
4. I standardized all redirects to `stderr` with `>&2`.
5. I fixed a few more `if`/`for` loops broken into two lines, as well as more POSIX tests ( `[ ]` ) and some unquoted variables.
6. When encountered ( more work to be done here ) I moved the declaration of `local` variables to the top of the function.
7. In `check_tags.sh` I removed unnecessary sub-shells. Since we only care about the return code of the `contains` function, we can read it directly.
```bash
if [[ -z $( contains "arm64.deb" ) ]]; then
```
becomes
```bash
if contains "arm64.deb"; then
````
8. Last and not least, in `icons/build_icons.sh` I refactored the entire `build_windows_types` function. Now the file is being checked with the following function:
```bash
create_image() {
    local PATH SIZE COLOR SVG LOGO GRAVITY

    PATH="$1"
    SIZE="$2"
    COLOR="$3"
    SVG="$4"
    GRAVITY="$5"

  if [[ ! -f "${PATH}" ]]; then
    convert -size "${SIZE}" "${COLOR}":"${PATH}"
    rsvg-convert -w "${SVG}" -h "${SVG}" "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"
    composite -gravity "${GRAVITY}" "code_logo.png" "${PATH}" "${PATH}"
  fi
}
```
With each variation being called:
```bash
create_image "${SRC_PREFIX}src/${QUALITY}/resources/win32/code_70x70.png" "70x70" "canvas:transparent PNG32" "45" "center"
```
This removed 60 lines of repeated code. There are several other scripts that could benefit from this type of refactoring ( repeated `if`/`fi` blocks ) that I'll tackle next week if you like this.

### CODE REVIEW

**Comment by daiyam** [comment]
`${var,,}` doesn't work on macOS...

The new function `create_image()` will need to be reworked. I will look into it.

**Comment by zpiatt** [comment]
> `${var,,}` doesn't work on macOS...
> 
> The new function `create_image()` will need to be reworked. I will look into it.

I will switch back to the `awk` command. 

Here's my other idea for refactoring that function (does away with the new function):
```bash
build_windows_types() { # {{{
  local FILEPATH SIZE COLOR SVG GRAVITY

  FILEPATH=(
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/code_70x70.png"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/code_150x150.png"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-100.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-125.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-150.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-175.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-200.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-225.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-big-250.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-100.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-125.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-150.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-175.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-200.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-225.bmp"
    "${SRC_PREFIX}src/${QUALITY}/resources/win32/inno-small-250.bmp"
    "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-banner.bmp"
    "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-dialog.bmp"
    )

  SIZE=(
    "70x70" "150x150" "164x314" "192x386" "246x459" "273x556"
    "328x604" "355x700" "410x797" "55x55" "64x68" "83x80" "92x97"
    "110x106" "119x123" "138x140" "493x58" "493x312"
    )

  COLOR=(
    "canvas:transparent PNG32" "canvas:transparent PNG32" "xc:white"
    "xc:white" "xc:white" "xc:white" "xc:white" "xc:white" "xc:white"
    "xc:white" "xc:white" "xc:white" "xc:white" "xc:white" "xc:white"
    "xc:white" "xc:white" "xc:white"
    )

  SVG=(
    "126" "147" "190" "211" "255" "273" "317" "44" "52"
    "63" "76" "86" "103" "116" "50" "120" "45" "64"
    )

  GRAVITY=(
    "center" "center" "center" "center" "center" "center" "center" "center"
    "center" "center" "center" "center" "center" "center" "+438+6" "+22+152"
  )

  mkdir -p "${SRC_PREFIX}src/${QUALITY}/resources/win32"

  rsvg-convert -b "#F5F6F7" -w 64 -h 64 "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"

  for file in "${VSCODE_PREFIX}"vscode/resources/win32/*.ico; do
    if [[ -f "${file}" ]]; then
      name=$(basename "${file}" '.ico')

      if [[ "${name}" != 'code' ]] && [[ ! -f "${SRC_PREFIX}src/${QUALITY}/resources/win32/${name}.ico" ]]; then
        icotool -x -w 256 "${file}"

        composite -geometry +150+185 "code_logo.png" "${name}_9_256x256x32.png" "${name}.png"

        convert "${name}.png" -define icon:auto-resize=256,128,96,64,48,32,24,20,16 "${SRC_PREFIX}src/${QUALITY}/resources/win32/${name}.ico"

        rm "${name}_9_256x256x32.png" "${name}.png"
      fi
    fi
  done

  for i in "${!FILEPATH[@]}"; do
    if [[ ! -f "${FILEPATH[i]}" ]]; then
      convert -size "${SIZE[i]}" "${COLOR[i]}":"${FILEPATH[i]}"
      rsvg-convert -w "${SVG[i]}" -h "${SVG[i]}" "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"
      composite -gravity "${GRAVITY[i]}" "code_logo.png" "${FILEPATH[i]}" "${FILEPATH[i]}"
    fi
  done

  rm code_logo.png
} # }}}
```

**Comment by zpiatt** [comment]
Reverted `awk` commands. Pushed new `build_windows_type` using arrays in lieu of new function. If you like this, it could save a lot of repeated code in `check_tags.sh`. Let me know!

File: update_version.sh [hunk 1]
```diff
@@ -117,7 +117,7 @@ updateLatestVersion() {
 # thank you https://www.vinaygopinath.me/blog/tech/commit-to-master-branch-on-github-using-travis-ci/
 git clone "https://github.com/${VERSIONS_REPOSITORY}.git"
 cd "${REPOSITORY_NAME}" || { echo "'${REPOSITORY_NAME}' dir not found"; exit 1; }
-git config user.email "$( echo "${GITHUB_USERNAME}" | awk '{print tolower($0)}' )-ci@not-real.com"
+git config user.email "${GITHUB_USERNAME,,}-ci@not-real.com"
```
Comment (daiyam) [inline update_version.sh:120]: please revert to awk

**Comment by daiyam** [comment]
The new `build_windows_types()` isn't friendly. Please revert to your previous version. I will give you an updated version (renamed variables and an additional variable)

File: update_version.sh (hunk 1 above)
Comment (zpiatt) [reply update_version.sh:120]: Done, I also caught this in `check_tag.sh` and `update_insider`.

**Comment by zpiatt** [comment]
> The new `build_windows_types()` isn't friendly. Please revert to your previous version. I will give you an updated version (renamed variables and an additional variable)

Reverted, do you want me to push it, or stand by for your changes?

**Comment by daiyam** [comment]
Here my updated version:
```
build_windows_type() {
  local PATH IMG_SIZE IMG_BG_COLOR LOGO_SIZE GRAVITY

  PATH="$1"
  IMG_SIZE="$2"
  IMG_BG_COLOR="$3"
  LOGO_SIZE="$4"
  GRAVITY="$5"

  if [[ ! -f "${PATH}" ]]; then
    if [[ "${PATH##*.}" == "png" ]]; then
      convert -size "${IMG_SIZE}" "${IMG_BG_COLOR}" PNG32:"${PATH}"
    else
      convert -size "${IMG_SIZE}" "${IMG_BG_COLOR}" "${PATH}"
    fi

    rsvg-convert -w "${LOGO_SIZE}" -h "${LOGO_SIZE}" "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"

    composite -gravity "${GRAVITY}" "code_logo.png" "${PATH}" "${PATH}"
  fi
}
```

I've renamed the variables to make the more clear about what they do.
I didn't liked the `PNG32` included in the `COLOR`, so now it's detect if the target is an PNG, then make sure that imagemagick use a 32 bits png (so when merging the svg, the logo looks correct).

But not sure why, I get `convert: command not found` while `convert` is found when outside the if... What do I miss???

**Comment by zpiatt** [comment]
> Here my updated version:
> 
> ```
> build_windows_type() {
>   local PATH IMG_SIZE IMG_BG_COLOR LOGO_SIZE GRAVITY
> 
>   PATH="$1"
>   IMG_SIZE="$2"
>   IMG_BG_COLOR="$3"
>   LOGO_SIZE="$4"
>   GRAVITY="$5"
> 
>   if [[ ! -f "${PATH}" ]]; then
>     if [[ "${PATH##*.}" == "png" ]]; then
>       convert -size "${IMG_SIZE}" "${IMG_BG_COLOR}" PNG32:"${PATH}"
>     else
>       convert -size "${IMG_SIZE}" "${IMG_BG_COLOR}" "${PATH}"
>     fi
> 
>     rsvg-convert -w "${LOGO_SIZE}" -h "${LOGO_SIZE}" "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"
> 
>     composite -gravity "${GRAVITY}" "code_logo.png" "${PATH}" "${PATH}"
>   fi
> }
> ```
> 
> I've renamed the variables to make the more clear about what they do. I didn't liked the `PNG32` included in the `COLOR`, so now it's detect if the target is an PNG, then make sure that imagemagick use a 32 bits png (so when merging the svg, the logo looks correct).
> 
> But not sure why, I get `convert: command not found` while `convert` is found when outside the if... What do I miss???

Not sure if it's related, but I think it was a mistake to use "PATH". I'm going to switch it to "FILE_PATH" to ensure it doesn't expand to the environmental variable.

**Comment by daiyam** [comment]
> Not sure if it's related, but I think it was a mistake to use "PATH". I'm going to switch it to "FILE_PATH" to ensure it doesn't expand to the environmental variable.

Yep, it was!

File: icons/build_icons.sh [hunk 2]
```diff
@@ -121,99 +143,24 @@ build_windows_types() { # {{{
...
-    convert -size 493x58 xc:white "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-banner.bmp"
-    rsvg-convert -w 50 -h 50 "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"
-    composite -geometry +438+6 "code_logo.png" "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-banner.bmp" "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-banner.bmp"
-  fi
-  if [[ ! -f "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-dialog.bmp" ]]; then
-    convert -size 493x312 xc:white "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-dialog.bmp"
-    rsvg-convert -w 120 -h 120 "icons/${QUALITY}/codium_cnl.svg" -o "code_logo.png"
-    composite -geometry +22+152 "code_logo.png" "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-dialog.bmp" "${SRC_PREFIX}build/windows/msi/resources/${QUALITY}/wix-dialog.bmp"
-  fi
+  create_image "${SRC_PREFIX}src/${QUALITY}/resources/win32/code_70x70.png" "70x70" "canvas:transparent" "45" "center"
```
Comment (daiyam) [inline icons/build_icons.sh:146]: Can you rename it to `build_windows_type`? Then we should be good.

Comment (zpiatt) [reply icons/build_icons.sh:146]: Yes, sorry completely missed that part.
Two other quick questions. 
1. In `build.sh` I want to make sure the intent of line 33:
`find "../VSCode-linux-${VSCODE_ARCH}" -exec touch {} \;` is to update an existing file. I added `-c` to ensure it doesn't accidentally create an empty file. ( `find "../VSCode-linux-${VSCODE_ARCH}" -print0 | xargs -0 touch -c` ).
2. Is this expansion valid in macos: `cp package.json{,.bak}` or should I revert to `cp package.json package.json.bak`?

Comment (daiyam) [reply icons/build_icons.sh:146]: 1. It's to make sure that the file has the current date instead to have them in 1970...
2. Yes

Comment (daiyam) [reply icons/build_icons.sh:146]: macOS is still using Bash 3 will the lower casing is available in Bash 4.

File: build/windows/msi/build.sh [hunk 3]
```diff
@@ -49,7 +49,7 @@ fi
 sed -i "s|@@PRODUCT_UPGRADE_CODE@@|${PRODUCT_UPGRADE_CODE}|g" .\\includes\\vscodium-variables.wxi
 sed -i "s|@@PRODUCT_NAME@@|${PRODUCT_NAME}|g" .\\vscodium.xsl
 
-find i18n -name '*.wxl' -exec sed -i "s|@@PRODUCT_NAME@@|${PRODUCT_NAME}|g" {} \;
+find i18n -name '*.wxl' print0 | xargs -0 sed -i "s|@@PRODUCT_NAME@@|${PRODUCT_NAME}|g"
```
Comment (zpiatt) [inline build/windows/msi/build.sh:52]: This typo is causing the failure on the windows build. Should be `-print0`

Comment (zpiatt) [reply build/windows/msi/build.sh:52]: I don't see anything else causing the failures. Re-pushing with this fixed.

**Comment by daiyam** [comment]
Is it ok to merge?

**Comment by zpiatt** [comment]
> Is it ok to merge?

I believe so, passing all workflows now, and no errors found from `shellcheck` linting.

**Comment by daiyam** [comment]
@zpiatt Thanks again!

**Comment by daiyam** [comment]
I'm reverting the 7. since it doesn't correctly detect the assets...

**Comment by daiyam** [comment]
I've also fixed a typo in #1577

**Pull request merged**